### 1.Python Program to Check if a Number is Positive, Negative or 0

In [3]:
num = float(input("Enter a number: "))
if num > 0:
   print("Positive number")
elif num == 0:
   print("Zero")
else:
   print("Negative number")


Enter a number: 0
Zero


 > This program first prompts the user to enter a number using the input() function. The input is then converted to a floating point number using the float() function. Next, the program uses an if-else statement to check if the number is greater than 0, equal to 0, or less than 0. Depending on the outcome of this check, the program prints "Positive number", "Zero", or "Negative number" respectively.

### 2.Python Program to Check if a Number is Odd or Even

In [4]:
num = int(input("Enter a number: "))
if num % 2 == 0:
   print("Even")
else:
   print("Odd")


Enter a number: 78
Even


> This program first prompts the user to enter a number using the input() function. The input is then converted to an integer using the int() function. Next, the program uses the modulo operator (%) to check the remainder when the number is divided by 2. If the remainder is 0, it means the number is even and the program prints "Even". If the remainder is not 0, it means the number is odd and the program prints "Odd".

### 3.Python Program to Check Leap Year

In [6]:
year = int(input("Enter a year: "))

if (year % 4) == 0:
   if (year % 100) == 0:
       if (year % 400) == 0:
           print("{0} is a leap year".format(year))
       else:
           print("{0} is not a leap year".format(year))
   else:
       print("{0} is a leap year".format(year))
else:
   print("{0} is not a leap year".format(year))


Enter a year: 2023
2023 is not a leap year


>This program first prompts the user to enter a year using the input() function. The input is then converted to an integer using the int() function. Next, the program uses a series of if-else statements to check if the year is a leap year.

A leap year is a year that is exactly divisible by 4, except for end-of-century years (years that end in 00), which must be exactly divisible by 400.

In this program, first, it checks whether the year is exactly divisible by 4 or not, if yes, it checks whether the year is exactly divisible by 100 or not, if yes it checks whether the year is exactly divisible by 400 or not. If the year passes all these condition it will be a leap year otherwise not.

The program prints the appropriate message, either "year is a leap year" or "year is not a leap year" depending on the outcome of the checks.

### 4. Python Program to Find the Largest Among Three Numbers



In [7]:
num1 = float(input("Enter first number: "))
num2 = float(input("Enter second number: "))
num3 = float(input("Enter third number: "))

if (num1 >= num2) and (num1 >= num3):
   largest = num1
elif (num2 >= num1) and (num2 >= num3):
   largest = num2
else:
   largest = num3

print("The largest number is", largest)


Enter first number: 398
Enter second number: 509
Enter third number: 570
The largest number is 570.0


> This program prompts the user to enter three numbers using the input() function, which are then converted to floating point numbers using the float() function. Next, the program uses an if-elif-else statement to check which of the three numbers is the largest. The largest number is then stored in the variable largest and is printed to the screen using the print() function.

In [8]:
# 4.You can also use the max() function to find the largest among three numbers

num1 = float(input("Enter first number: "))
num2 = float(input("Enter second number: "))
num3 = float(input("Enter third number: "))

largest = max(num1, num2, num3)

print("The largest number is", largest)



Enter first number: 678
Enter second number: 980
Enter third number: 567
The largest number is 980.0


> This program also prompts the user to enter three numbers using the input() function, which are then converted to floating point numbers using the float() function. Next, the program uses the max() function to find the largest among three numbers, which is then stored in the variable largest and is printed to the screen using the print() function.

### 5.Python Program to Check Prime Number

In [9]:
num = int(input("Enter a number: "))

if num > 1:
    for i in range(2,num):
        if (num % i) == 0:
            print(num,"is not a prime number")
            break
    else:
        print(num,"is a prime number")
else:
   print(num,"is not a prime number")


Enter a number: 95
95 is not a prime number


> This program first prompts the user to enter a number using the input() function, which is then converted to an integer using the int() function. Next, the program uses a for loop to iterate over a range of numbers starting from 2 and ending at the number entered by the user. Inside the loop, it checks if the entered number is divisible by the current iteration number using the modulo operator (%), if it is divisible that means it's not a prime number and the program prints the message and breaks the loop, otherwise, the for loop will complete and the program will print the message that this number is prime. If the number is less than or equal to 1, it's not a prime number so the program prints the message accordingly.

In [10]:
# You can also use the sqrt() function to optimize 
# the above program, which will check the divisibility only till the square root of the number.

import math
num = int(input("Enter a number: "))
if num > 1:
    for i in range(2, int(math.sqrt(num)) + 1):
        if (num % i) == 0:
            print(num,"is not a prime number")
            break
    else:
        print(num,"is a prime number")
else:
   print(num,"is not a prime number")


Enter a number: 56
56 is not a prime number


> This program also works in the same way as the previous one, but it only checks for divisibility till the square root of the entered number, which makes the program more efficient

### 6.Python Program to Print all Prime Numbers in an Interval



In [14]:
start = int(input("Enter the start of the interval: "))
end = int(input("Enter the end of the interval: "))

print("Prime numbers between", start, "and", end, "are:")

for num in range(start, end + 1):
   if num > 1:
       for i in range(2,num):
           if (num % i) == 0:
               break
       else:
           print(num)


Enter the start of the interval: 34
Enter the end of the interval: 70
Prime numbers between 34 and 70 are:
37
41
43
47
53
59
61
67


> This program prompts the user to enter the start and end of the interval using the input() function, which are then converted to integers using the int() function. Next, the program uses a nested for loop to iterate over the numbers within the interval. The outer loop iterates over the numbers in the interval, while the inner loop checks if each number is prime.

In [15]:
import math
start = int(input("Enter the start of the interval: "))
end = int(input("Enter the end of the interval: "))

print("Prime numbers between", start, "and", end, "are:")

for num in range(start, end + 1):
   if num > 1:
       for i in range(2, int(math.sqrt(num)) + 1):
           if (num % i) == 0:
               break
       else:
           print(num)


Enter the start of the interval: 2
Enter the end of the interval: 10
Prime numbers between 2 and 10 are:
2
3
5
7


> This program also works in the same way as the previous one, but it only checks for divisibility till the square root of the entered number, which makes the program more efficient.

### 7.Python Program to Find the Factorial of a Number

In [16]:
num = int(input("Enter a number: "))

factorial = 1

if num < 0:
   print("Sorry, factorial does not exist for negative numbers")
elif num == 0:
   print("The factorial of 0 is 1")
else:
   for i in range(1,num + 1):
       factorial = factorial*i
   print("The factorial of",num,"is",factorial)


Enter a number: 6
The factorial of 6 is 720


> This program prompts the user to enter a number using the input() function, which is then converted to an integer using the int() function. Next, the program uses a for loop to iterate over a range of numbers starting from 1 and ending at the number entered by the user. Inside the loop, it multiply each iteration number with the previous iteration result. If the number is less than 0, it's not defined so the program prints the message accordingly. If the number is 0, the factorial is 1, so the program prints the message accordingly. Otherwise, the program prints the final factorial value.

In [17]:
# You can also use the recursion to find the factorial of a number

def factorial(n):
    if n == 0:
        return 1
    else:
        return n * factorial(n-1)

num = int(input("Enter a number: "))

if num < 0:
    print("Sorry, factorial does not exist for negative numbers")
else:
    print("The factorial of",num,"is",factorial(num))


Enter a number: 5
The factorial of 5 is 120


> This program uses a recursive function factorial() to find the factorial of a number. The base case is when the number is 0, in that case the function returns 1. In all other cases, the function calls itself with the number decremented by 1 and multiplies it with the original number. The final result is returned and printed on the screen.

### 8.Python Program to Display the multiplication Table

In [18]:
num = int(input("Enter a number to display its multiplication table: "))

for i in range(1, 11):
   print(num, 'x', i, '=', num*i)


Enter a number to display its multiplication table: 5
5 x 1 = 5
5 x 2 = 10
5 x 3 = 15
5 x 4 = 20
5 x 5 = 25
5 x 6 = 30
5 x 7 = 35
5 x 8 = 40
5 x 9 = 45
5 x 10 = 50


> This program prompts the user to enter a number using the input() function, which is then converted to an integer using the int() function. Next, the program uses a for loop to iterate over a range of numbers from 1 to 11. Inside the loop, it multiplies the entered number with the current iteration number, and prints the multiplication table using the print() function.

In [21]:
# You can also display the multiplication table up to a 
# certain number by changing the range of the for loop
num = int(input("Enter a number to display its multiplication table: "))
n = int(input("Enter the number of terms to display the multiplication table: "))

for i in range(1, n+1):
   print(num, 'x', i, '=', num*i)


Enter a number to display its multiplication table: 1
Enter the number of terms to display the multiplication table: 10
1 x 1 = 1
1 x 2 = 2
1 x 3 = 3
1 x 4 = 4
1 x 5 = 5
1 x 6 = 6
1 x 7 = 7
1 x 8 = 8
1 x 9 = 9
1 x 10 = 10


> This program also prompts the user to enter a number and the number of terms to display the multiplication table. The program then uses a for loop to iterate over a range of numbers from 1 to the number of terms entered by the user. Inside the loop, it multiplies the entered number with the current iteration number and prints the multiplication table using the print() function.

### 9.Python Program to Print the Fibonacci sequence

In [23]:
nterms = int(input("How many terms? "))

n1, n2 = 0, 1
count = 0

if nterms <= 0:
   print("Please enter a positive integer")
elif nterms == 1:
   print("Fibonacci sequence upto",nterms,":")
   print(n1)
else:
   print("Fibonacci sequence:")
   while count < nterms:
       print(n1)
       nth = n1 + n2
       n1 = n2
       n2 = nth
       count += 1


How many terms? 6
Fibonacci sequence:
0
1
1
2
3
5


> This program prompts the user to enter the number of terms in the Fibonacci sequence using the input() function, which is then converted to an integer using the int() function. Next, the program uses a while loop to iterate over the terms in the sequence. The loop starts with 0 and 1 as the first two terms and continues until the number of terms entered by the user. Inside the loop, it adds the previous two terms to get the next term and reassigns the variables, at last it prints the term.

In [24]:
# You can also use recursion to print the Fibonacci sequence.
def fibonacci(n):
    if n <= 0:
        return None
    elif n == 1:
        return 0
    elif n == 2:
        return 1
    else:
        return fibonacci(n-1) + fibonacci(n-2)

nterms = int(input("How many terms? "))

if nterms <= 0:
    print("Please enter a positive integer")
else:
    print("Fibonacci sequence:")
    for i in range(1, nterms + 1):
        print(fibonacci(i))


How many terms? 5
Fibonacci sequence:
0
1
1
2
3


> This program uses a recursive function fibonacci() to find the Fibonacci sequence. The base cases are when the number is less than or equal to 0, it returns None, when the number is 1, it returns 0 and when the number is 2, it returns 1. In all other cases, the function calls itself twice with the number decremented by 1 and 2 respectively and adds the results. The final result is returned and printed on the screen in a loop with the range of 1 to the number of terms entered by the user.

### 10.Python Program to Check Armstrong Number

In [30]:
num = int(input("Enter a number: "))

sum = 0
temp = num

while temp > 0:
   digit = temp % 10
   sum += digit ** 3
   temp //= 1024

if num == sum:
   print(num,"is an Armstrong number")
else:
   print(num,"is not an Armstrong number")


Enter a number: 371
371 is not an Armstrong number


> An Armstrong number is a number that is equal to the sum of its own digits each raised to the power of the number of digits.

This program prompts the user to enter a number using the input() function, which is then converted to an integer using the int() function. Next, the program uses a while loop to iterate over the digits of the number. Inside the loop, it finds the last digit of the number using the modulo operator(%), raise it to the power of the number of digits and add it to the sum, then it integer division the number by 10 to get the next digit.

After the loop, if the sum is equal to the original number, it is an Armstrong number and the program prints the message accordingly, otherwise, it's not an Armstrong number so the program prints the message accordingly.

### 11.Python Program to Find Armstrong Number in an Interval

In [33]:
start = int(input("Enter the start of the interval: "))
end = int(input("Enter the end of the interval: "))

print("Armstrong numbers between", start, "and", end, "are:")

for num in range(start, end + 1):
    sum = 0
    temp = num
    n = len(str(num))
    while temp > 0:
        digit = temp % 10
        sum += digit ** n
        temp //= 10
    if num == sum:
       print(num)


Enter the start of the interval: 120
Enter the end of the interval: 500
Armstrong numbers between 120 and 500 are:
153
370
371
407


> An Armstrong number of three digits is an integer such that the sum of the cubes of its digits is equal to the number itself. For example, 371 is an Armstrong number since 3**3 + 7**3 + 1**3 = 371. Write a program to find all Armstrong number in the range of 0 and 999.

# 12.Python Program to Find the Sum of Natural Numbers

In [3]:
n = int(input("Enter the number: "))

# initialize sum
sum = 0

# use while loop to iterate through natural numbers
i = 1
while i <= n:
    sum += i
    i += 1

print("The sum of natural numbers up to", n, "is", sum)


Enter the number: 10
The sum of natural numbers up to 10 is 55


This program prompts the user to enter a number, and then uses a while loop to iterate through the natural numbers up to that number, adding each one to a running total. The final sum is then printed to the screen.

In [4]:
# Another way using for loop
n = int(input("Enter the number: "))

# initialize sum
sum = 0

# use for loop to iterate through natural numbers
for i in range(1,n+1):
    sum += i

print("The sum of natural numbers up to", n, "is", sum)


Enter the number: 23
The sum of natural numbers up to 23 is 276


# Practice Assignments : Files & Exception Handling

# 1.Create a File self.txt and write yourself introduction to it.

You can use the open() function to create a new file or open an existing file in Python. The function takes two arguments: the name of the file, and the mode in which the file should be opened. To create a new file and write to it, you should use the 'w' mode. Here's an example:



In [ ]:
# Open a file named 'self.txt' in write mode
file = open('self.txt', 'w')

# Write your introduction to the file
introduction = "Hello, I am an AI model trained by OpenAI. I am here to assist with any questions you may have."
file.write(introduction)

# Close the file
file.close()


In [ ]:
# Write your introduction to the file
introduction = "Hello, I am an AI model trained by OpenAI. I am here to assist with any questions you may have."
with open('self.txt', 'w') as file:
    file.write(introduction)


# 2.Append your phone number & address in the file self.txt which you had already created during last exercise PA : FEH 1 

To append additional information to an existing file, you should open the file in 'append' mode ('a') rather than write mode ('w'). You can use the write() method to add new data to the end of the file. Here's an example:

In [ ]:
# Open the file 'self.txt' in append mode
file = open('self.txt', 'a')

# Append your phone number and address to the file
file.write("\nMy phone number is 555-555-5555.")
file.write("\nMy address is 123 Main St.")

# Close the file
file.close()


In [ ]:
# Append your phone number and address to the file
with open('self.txt', 'a') as file:
    file.write("\nMy phone number is 555-555-5555.")
    file.write("\nMy address is 123 Main St.")


# 3.Read your entire file self.txt file that you had created during PA : FEH 1 exercise

In [ ]:
# Open the file 'self.txt' in read mode
file = open('self.txt', 'r')

# Read the entire file
contents = file.read()

# Print the contents of the file
print(contents)

# Close the file
file.close()


This code will open the file 'self.txt' in read mode, read the entire contents of the file and print it to the screen.

You can also use with open statement which will automatically close the file for you

In [ ]:
# Read the entire file
with open('self.txt', 'r') as file:
    contents = file.read()
    print(contents)


# 4.Check the existence of two files using exception handling



You can use try-except blocks to handle exceptions in Python, including the FileNotFoundError exception that is raised when trying to open a non-existent file. Here is an example of how you can check the existence of two files using exception handling:
    

In [ ]:
filenames = ['file1.txt', 'file2.txt']

for filename in filenames:
    try:
        with open(filename, 'r') as file:
            contents = file.read()
            print(f"{filename} exists and its content is: {contents}")
    except FileNotFoundError:
        print(f"{filename} does not exist.")


In this example, the program tries to open each file in the 'filenames' list with with open statement which automatically closes the file. If the file exists, it will read the contents of the file and print it to the screen. If the file does not exist, a FileNotFoundError exception is raised and the program will print a message saying that the file does not exist.

You can also use the os.path.exists(path) method to check if a file exists or not and use this method in an if-else statement to handle the scenario.

In [ ]:
import os
filenames = ['file1.txt', 'file2.txt']

for filename in filenames:
    if os.path.exists(filename):
        with open(filename, 'r') as file:
            contents = file.read()
            print(f"{filename} exists and its content is: {contents}")
    else:
        print(f"{filename} does not exist.")


# 5.Create a new file using X mode in tryu exception block

In [ ]:
file_name = "new_file.txt"
try:
    with open(file_name, 'x') as file:
        file.write("This is a new file.")
    print(f"{file_name} created successfully.")
except FileExistsError:
    print(f"{file_name} already exists.")


This code will try to create a new file named 'new_file.txt' in the current directory with with open statement which automatically closes the file. If the file already exists, the FileExistsError exception is raised and the program will print a message saying that the file already exists. If the file does not exist, it will create the file and write the given string to it and print a message saying that the file has been created successfully.

You can also use the os.path.exists(path) method to check if a file exists or not and use this method in an if-else statement to handle the scenario.

In [ ]:
import os
file_name = "new_file.txt"
if os.path.exists(file_name):
    print(f"{file_name} already exists.")
else:
    with open(file_name, 'x') as file:
        file.write("This is a new file.")
    print(f"{file_name} created successfully.")


Both examples will check the existence of the files before creating a new one and raise FileExistsError exception if the file already exists. If the file does not exist it will create a new file and write the given string to it.

# 5.Write a function that asks for an integer and prints the square of it. Use a while loop with a try, except, else block to account for incorrect inputs.



In [5]:
def ask_for_int():
    while True:
        try:
            number = int(input("Please enter an integer: "))
        except ValueError:
            print("Invalid input. Please enter an integer.")
            continue
        else:
            print("The square of the number is:", number**2)
            break


This function uses a while loop to continually prompt the user for an integer input until a valid input is provided. Inside the loop, it uses a try block to attempt to convert the input to an integer using the int() function. If the input is not a valid integer (e.g. if the user enters a string or a decimal), a ValueError exception is raised and the except block is executed, which prints an error message and continues the loop, prompting the user for another input.

If the input is a valid integer, the else block is executed, which calculates the square of the input number, prints the result and breaks the loop.

This function will keep asking the user for an integer input until the user enters an integer and it will print the square of the entered integer.





# 7.Write a python program to create a exception class in 1. Arithmetic Error 2. Import Error 3. Index Error



In [6]:
class ArithmeticError(Exception):
    def __init__(self, message):
        self.message = message

try:
    num1 = int(input("Enter first number: "))
    num2 = int(input("Enter second number: "))
    if num2 == 0:
        raise ArithmeticError("division by zero")
    print("Result: ", num1/num2)
except ArithmeticError as e:
    print(e)


Enter first number: 3
Enter second number: 2
Result:  1.5


This code creates a custom exception class called 'ArithmeticError' that inherits from the built-in 'Exception' class. When the user inputs 0 as the second number and the program tries to divide the first number by 0, the code raises the custom exception with an error message "division by zero" and the except block will catch the exception and print the error message.

In [7]:
# ImportError:
class ImportError(Exception):
    def __init__(self, message):
        self.message = message

try:
    import module_that_doesnt_exist
except ImportError as e:
    print("ImportError:", e)


ModuleNotFoundError: No module named 'module_that_doesnt_exist'

This code creates a custom exception class called 'ImportError' that inherits from the built-in 'Exception' class. When the program tries to import a module that doesn't exist, the code raises the custom exception with an error message and the except block will catch the exception and print the error message.

In [8]:
# IndexError:
class IndexError(Exception):
    def __init__(self, message):
        self.message = message

try:
    lst = [1, 2, 3]
    print(lst[3])
except IndexError as e:
    print("IndexError:", e)


IndexError: list index out of range

This code creates a custom exception class called 'IndexError' that inherits from the built-in 'Exception' class. When the program tries to access an index that is out of range of the list, the code raises the custom exception with an error message and the except block will catch the exception and print the error message.

Please note that it is not recommended to create custom exception classes with the same name as built-in exceptions, it is better to use a different naming convention.





# 7.Write a python program to create a exception class in 1. Arithmetic Error 2. Import Error 3. Index Error



In [9]:
class MyArithmeticError(Exception):
    def __init__(self, message):
        self.message = message

try:
    num1 = int(input("Enter first number: "))
    num2 = int(input("Enter second number: "))
    if num2 == 0:
        raise MyArithmeticError("division by zero")
    print("Result: ", num1/num2)
except MyArithmeticError as e:
    print(e)


Enter first number: 3
Enter second number: 6
Result:  0.5


This code creates a custom exception class called 'MyArithmeticError' that inherits from the built-in 'Exception' class. When the user inputs 0 as the second number and the program tries to divide the first number by 0, the code raises the custom exception with an error message "division by zero" and the except block will catch the exception and print the error message.

In [ ]:
# ImportError:
class MyImportError(Exception):
    def __init__(self, message):
        self.message = message

try:
    import module_that_doesnt_exist
except MyImportError as e:
    print("MyImportError:", e)


This code creates a custom exception class called 'MyImportError' that inherits from the built-in 'Exception' class. When the program tries to import a module that doesn't exist, the code raises the custom exception with an error message and the except block will catch the exception and print the error message.

In [10]:
# IndexError:
class MyIndexError(Exception):
    def __init__(self, message):
        self.message = message

try:
    lst = [1, 2, 3]
    print(lst[3])
except MyIndexError as e:
    print("MyIndexError:", e)


IndexError: list index out of range

This code creates a custom exception class called 'MyIndexError' that inherits from the built-in 'Exception' class. When the program tries to access an index that is out of range of the list, the code raises the custom exception with an error message and the except block will catch the exception and print the error message.

Please note that it is not recommended to create custom exception classes with the same name as built-in exceptions, it is better to use a different naming convention.





# 8.Write a python program of getting integer input from user. If user does not provide input, program should keep asking for it.



In [11]:
while True:
    try:
        number = int(input("Please enter an integer: "))
        break
    except ValueError:
        print("Invalid input. Please enter an integer.")


Please enter an integer: 12


This program uses a while loop to continually prompt the user for an integer input. Inside the loop, it uses a try block to attempt to convert the input to an integer using the int() function. If the input is not a valid integer (e.g. if the user enters a string or a decimal), a ValueError exception is raised and the except block is executed, which prints an error message and continues the loop, prompting the user for another input.

If the input is a valid integer, the try block will complete successfully and the program will exit the loop.

You can also use the input() function to get the input and check if it is a number or not using isdigit() function and use an if-else statement to handle the scenario.

In [13]:
while True:
    number = input("Please enter an integer: ")
    if number.isdigit():
        number = int(number)
        break
    else:
        print("Invalid input. Please enter an integer.")


Please enter an integer: -1
Invalid input. Please enter an integer.
Please enter an integer: 2


This program is also doing the same thing but this time it uses the isdigit() method to check if the input is an integer or not, if the input is not an integer it will keep asking for it until it gets an integer input from the user.

# 9.Write a python program to open .txt document using function and Exception handling

In [ ]:
def open_file(file_name):
    try:
        with open(file_name, 'r') as file:
            return file.read()
    except FileNotFoundError:
        print(f"{file_name} not found.")
    except:
        print("An error occurred.")

file_name = input("Enter the file name: ")
print(open_file(file_name))


In this example, the function open_file takes a file name as its parameter and tries to open it using the open function. If the file is successfully opened, the function returns its contents using the read() method. If a FileNotFoundError is encountered, the program prints a message indicating that the file was not found. The final except block catches any other errors that may occur. The input is taken from the user to provide the file name.





# 10.Handle the exception thrown by the code below by using try and except blocks. for i in ['a','b','c']:print(i**2)

In [ ]:
try:
    for i in ['a','b','c']:
        print(i**2)
except TypeError:
    print("TypeError: unsupported operand type(s) for ** or pow(): 'str' and 'int'")


The for loop iterates over the elements in the list ['a','b','c'] and tries to raise each element to the power of 2. However, since the elements in the list are strings, this operation will raise a TypeError (unsupported operand type(s) for ** or pow(): 'str' and 'int'). The try block contains the code that might raise the exception, and the except block contains the code that will be executed if an exception is raised. In this case, it prints the message "TypeError: unsupported operand type(s) for ** or pow(): 'str' and 'int'"

Note: The above example is useful when you are unaware of the input that your program is going to receive. But if you are aware of the input and it's a string then you can convert the input to int and then perform the operation.

#  Functions

# 1.Python Program To Display Powers of 2 Using Anonymous Function



In [18]:
n = int(input("Enter the number of powers of 2 to display: "))

powers_of_2 = map(lambda x: 2**x, range(n))

for i in powers_of_2:
    print(i)


Enter the number of powers of 2 to display: 5
1
2
4
8
16


In this example, the input is taken from the user which is the number of powers of 2 to be displayed. map function applies the lambda function on every element of the range of numbers. The lambda function takes one parameter x and raises 2 to the power of x. The result is stored in powers_of_2.
Then the for loop iterates through the powers_of_2 and prints the value of each element.

In [21]:
powers_of_2 = [2 ** i for i in range(n)]
# This will achieve the same result as the above code snippet.

# 2.Python Program to Find Numbers Divisible by Another Number

In [22]:
def divisible_numbers(n, divisor):
    return [x for x in range(n) if x % divisor == 0]

n = int(input("Enter the upper range for numbers: "))
divisor = int(input("Enter the divisor: "))

result = divisible_numbers(n, divisor)

if len(result) > 0:
    print(f"Numbers divisible by {divisor} between 0 and {n-1} are: ",result)
else:
    print(f"No numbers found between 0 and {n-1} that are divisible by {divisor}")


Enter the upper range for numbers: 4
Enter the divisor: 2
Numbers divisible by 2 between 0 and 3 are:  [0, 2]


This program takes two inputs from the user, the upper range of numbers and divisor.
The function divisible_numbers takes two parameters: n and divisor. It uses list comprehension to create a list of numbers between 0 and n-1 that are divisible by divisor. It then checks the length of the result, if it is greater than 0 then it prints the numbers else it prints no numbers are divisible.

In [ ]:
# result = list(filter(lambda x: x % divisor == 0, range(n)))
# This will filter out the numbers that are divisible by the divisor and return the same result.

# 3.Python Program to Convert Decimal to Binary, Octal and Hexadecimal

In [24]:
decimal = int(input("Enter a decimal number: "))

print("The binary representation of", decimal, "is", bin(decimal))
print("The octal representation of", decimal, "is", oct(decimal))
print("The hexadecimal representation of", decimal, "is", hex(decimal))


Enter a decimal number: 4
The binary representation of 4 is 0b100
The octal representation of 4 is 0o4
The hexadecimal representation of 4 is 0x4


In this example, the input is taken from the user which is a decimal number.
The bin() function is used to convert the decimal number to its binary representation, oct() function is used to convert the decimal number to its octal representation and hex() function is used to convert the decimal number to its hexadecimal representation.
All the above functions return the result in the form of string, so you can use string formatting to get the desired output.

In [25]:
print("The binary representation of {} is {}".format(decimal, bin(decimal)[2:]))
print("The octal representation of {} is {}".format(decimal, oct(decimal)[2:]))
print("The hexadecimal representation of {} is {}".format(decimal, hex(decimal)[2:]))


The binary representation of 4 is 100
The octal representation of 4 is 4
The hexadecimal representation of 4 is 4


The slice [2:] is used to remove the '0b' or '0o' or '0x' prefix from the string returned by the above functions.

# 4.Python Program to Find ASCII Value of Character

In [27]:
character = input("Enter a character: ")

if len(character) != 1:
    print("Please enter a single character.")
else:
    print("The ASCII value of", character, "is", ord(character))


Enter a character: 4
The ASCII value of 4 is 52


In this example, the input is taken from the user which is a single character. The ord() function is used to find the ASCII value of the character.
The ord() function returns the number representing the unicode code of a specified character.
The if-else statement is used to check if the user has entered a single character or not. If the user enters more than one character, it will print "Please enter a single character."

Note: ASCII is a subset of Unicode and the ASCII value of a character and its Unicode code point value are the same.

# 5.Python Program to Find HCF or GCD

In [28]:
import math

def find_hcf(a, b):
    return math.gcd(a, b)

num1 = int(input("Enter the first number: "))
num2 = int(input("Enter the second number: "))

print("The HCF/GCD of", num1, "and", num2, "is", find_hcf(num1, num2))


Enter the first number: 24
Enter the second number: 34
The HCF/GCD of 24 and 34 is 2


In this example, the input is taken from the user which are two numbers. The math.gcd() function is used to find the HCF of the two numbers. This function takes two integers as arguments and returns their greatest common divisor.
You can also use the below method for finding GCD:



# 6.Python Program to Find LCM by using functions

In [30]:
import math

def find_lcm(a, b):
    return (a*b)//math.gcd(a,b)

num1 = int(input("Enter the first number: "))
num2 = int(input("Enter the second number: "))

print("The LCM of", num1, "and", num2, "is", find_lcm(num1, num2))


Enter the first number: 12
Enter the second number: 24
The LCM of 12 and 24 is 24


In this example, the input is taken from the user which are two numbers. The math.gcd() function is used to find the GCD of the two numbers. Then the LCM is calculated by using the formula:



# 7.Python Program to Find the Factors of a Number by using Functions

In [31]:
def find_factors(n):
    factors = []
    for i in range(1,n+1):
        if n%i == 0:
            factors.append(i)
    return factors

num = int(input("Enter a number: "))

print("The factors of", num, "are", find_factors(num))


Enter a number: 12
The factors of 12 are [1, 2, 3, 4, 6, 12]


In this example, the input is taken from the user which is the number for which we want to find the factors.
The function find_factors takes one argument n and finds the factors of the number by using a for loop.
The for loop iterates from 1 to n+1 and checks if the current number is divisible by the input number, if it is divisible then it is added to the factors list.
The function returns the list of factors of the number.

Note: A factor of a number is an integer that can divide the number exactly without leaving a remainder.

# 8.Python Program to Make a Simple Calculator

In [33]:
def calculator(x, y, operation):
    if operation == '+':
        return x + y
    elif operation == '-':
        return x - y
    elif operation == '*':
        return x * y
    elif operation == '/':
        return x / y
    else:
        return "Invalid operation. Please use +, -, *, or /."

num1 = int(input("Enter the first number: "))
num2 = int(input("Enter the second number: "))
operation = input("Enter the operation (+, -, *, /): ")

print(calculator(num1, num2, operation))


Enter the first number: 23
Enter the second number: 24
Enter the operation (+, -, *, /): +
47


In this example, the input is taken from the user which are two numbers and the operation that the user wants to perform on the numbers.
The function calculator takes three arguments, x, y and operation. It checks which operation the user wants to perform and performs that operation on the two numbers.
It uses if-elif statement to check for different operations and return the result accordingly.
If the operation is not one of the four allowed operations, it returns an error message "Invalid operation. Please use +, -, *, or /."

# 9.Python Program to Find Sum of Natural Numbers Using Recursion



In [34]:
def sum_of_naturals(n):
    if n == 1:
        return 1
    else:
        return n + sum_of_naturals(n-1)

num = int(input("Enter a number: "))

print("The sum of the first", num, "natural numbers is", sum_of_naturals(num))


Enter a number: 12
The sum of the first 12 natural numbers is 78


In this example, the input is taken from the user which is the number upto which we want to find the sum of natural numbers.
The function sum_of_naturals takes one argument n and it will return the sum of first n natural numbers.
The function uses recursion, where it calls itself with an updated argument (n-1) until the base case is reached where n=1, then it will return 1.
The base case will be reached when n=1 and the function will return 1 and the recursion will end.

# 10.Python Program to Check Whether a String is Palindrome or Not

In [36]:
def is_palindrome(s):
    return s == s[::-1]

string = input("Enter a string: ")

if is_palindrome(string):
    print("The string is a palindrome.")
else:
    print("The string is not a palindrome.")


Enter a string: madam
The string is a palindrome.


In this example, the input is taken from the user which is the string that needs to be checked for palindrome.
The function is_palindrome takes one argument s and returns True if the string is a palindrome, False otherwise.
It uses slicing technique to reverse the string and check if the reversed string is equal to the original string or not.
Then the if-else statement is used to check if the string is palindrome or not and return the output accordingly.

# 11.Create an inner function to calculate the addition in the following way:

Create an outer function that will accept two parameters a and b. Create a function inside an outer function that will calculate the addition of a and b. At last, an outer function will add 5 into addition and return it



In [37]:
def outer_function(a, b):
    def inner_function():
        return a + b

    result = inner_function()
    return result + 5

x = int(input("Enter the first number: "))
y = int(input("Enter the second number: "))

print(outer_function(x, y))


Enter the first number: 5
Enter the second number: 5
15


In this example, the outer function outer_function takes two parameters a and b and defines an inner function inner_function within it. The inner function inner_function calculates the addition of a and b and returns it. The outer function then adds 5 to the result and returns the final result.
In the above code, the user is prompted to enter two numbers and the outer function is called with these numbers as arguments, which in turn calls the inner function to do the addition and adds 5 to it. The final result is printed.

Note: Inner functions can access the variables in the outer function, this is known as closure.